### PR TITLE
added support for docker images in tasks

### PR DESCRIPTION
**Pull request by no0dles**
Hi, really like your taskfile project and use it almost every day. I created a small proof of concept for running tasks inside docker containers. 

This feature mounts the current working directory into the container, executes the commands inside the container environment and forwards the log outputs. 

Personally I do not have a lot of go experience, maybe it could be done in a better way. Open for feedback.

### CODE REVIEW

File: internal/execext/exec.go [hunk 1]
```diff
@@ -3,6 +3,14 @@ package execext
...
 	"context"
 	"errors"
+	"fmt"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/term"
```
Comment (andreynering) [inline internal/execext/exec.go]: All imports from packages outside the standard library should be moved to the second block.

Comment (andreynering) [inline internal/execext/exec.go]: All imports from packages outside the standard library should be moved to the second block.

File: internal/execext/exec.go [hunk 2]
```diff
@@ -64,6 +77,86 @@ func RunCommand(ctx context.Context, opts *RunCommandOptions) error {
...
+		return err
+	}
+
+	pullReader, err := cli.ImagePull(context.Background(), image.String(), types.ImagePullOptions{})
+	if err != nil {
+		return err
+	}
+	defer pullReader.Close()
+
+	termFd, isTerminal := term.GetFdInfo(os.Stderr)
```
Comment (andreynering) [inline internal/execext/exec.go]: Should probably use `opts.Stderr` instead of `os.Stderr`.

File: internal/execext/exec.go [hunk 3]
```diff
@@ -64,6 +77,86 @@ func RunCommand(ctx context.Context, opts *RunCommandOptions) error {
...
+	err = jsonmessage.DisplayJSONMessagesStream(pullReader, os.Stderr, termFd, isTerminal, nil)
```
Comment (andreynering) [inline internal/execext/exec.go]: Same here

File: internal/execext/exec.go [hunk 4]
```diff
@@ -64,6 +77,86 @@ func RunCommand(ctx context.Context, opts *RunCommandOptions) error {
...
+		ShowStderr: true,
+		Follow:     true,
+	})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	go func() {
+		io.Copy(opts.Stdout, reader)
```
Comment (andreynering) [inline internal/execext/exec.go]: Question: how to handle any errors returned by `io.Copy`?

**Comment by no0dles** [comment]
Hey @andreynering 
Thank you very much for your feedback.

Agree with the --docker-image flag, only question would be if that flag applies to deps and sub-tasks.

Tried to fix your feedback and discovered another issue. Currently the context gets lost, because each step is in a separat container. This is bad for multiple use cases, for example this:

```
tasks:
  build:
    image: node:12
   cmds:
     - npm i -g typescript
     - tsc
```

Output: `in/sh: 1: tsc: not found`

**Comment by no0dles** [comment]
Another thing would be changing directory, currently the behaviour with and without an image is the same (see example below). But as soon as I would change the docker implementation to execute all the commands in one container, this would differ from the default way without an image.

```
tasks:
  test-pwd:
    dir: packages
    cmds:
      - pwd
      - cd ..
      - pwd
```

Output
```
test $ task test-pwd
task: pwd
/Users/pascal/test/dir
task: cd ..
task: pwd
/Users/pascal/test/dir

```

**Comment by no0dles** [comment]
what's the reason for closing this issue? never heard back from you.. would be sad to close this feature without any response

**Comment by andreynering** [comment]
Hi @no0dles,

Sorry, this was a mistake. GitHub closed many PRs automatically after merging #220

Reopening...

Also, I think you deserve an explanation on why I never came back to this PR: I was just busy, and focusing on features that will be released on v3 (hopefully very soon). I want to back to this at some point, but I can't promise when.

**Comment by orbatschow** [comment]
@andreynering  Are there any plans to merge this feature? (IMHO Docker images inside a Taskfile would be pretty awesome.)

**Comment by andreynering** [comment]
@orbatschow Sorry, but this may take some time.

I have been having little time to work on Task, and bug fixes are a priority at the moment.

**Comment by max-sixty** [comment]
This would be really useful for some tasks, where there are complex dependencies and a simple result. Some CI pipelines act similarly — select an image and a command — and this could be a cross-platform, abstract version of this. (Alternatively I'd be interested if there are alternatives that are better suited to this sort of task). Thanks.

**Comment by no0dles** [comment]
I am closing this PR since I created another [tool](https://github.com/no0dles/hammerkit) that solves that issue and it's easier for me to contribute there in a language I am familiar with. Thanks @andreynering

**Comment by andreynering** [comment]
Hi @no0dles,

You're did a great job on hammerkit, congrats! I saw other build tools inspired by Task out there, but I think this is the most interesting given the simplicity and Docker support. (I haven't tried it yet, though).

I was very in doubt on how an integration with Docker would look like and yours seems like great, simple and easy to understand approach. Hope you don't mind if I happen to take some inspiration from there 😉. That shouldn't happen too soon, though.

**Comment by no0dles** [comment]
@andreynering no worries, I did take some inspiration from taskfile as well 😄. Good luck, looking forward to a new release with this.

**Pull request closed without merging**